#!/usr/bin/python3

import pathlib
from Db import DBManipulator
from Ui import Ui_MalwareUploadDialog
from PyQt5.QtCore import *
from PyQt5.QtGui import *
from PyQt5.QtWidgets import *
from PyQt5 import QtCore, QtWidgets, QtGui
from Core import CoreException
from Core import CoreLoadConfig

class MalwareUploadDialog(QDialog,Ui_MalwareUploadDialog.Ui_Malware_Dialog):
    """
        Malware Upload Dialog.
    """
    trigger_messageBox = pyqtSignal("QString")
    trigger_success = pyqtSignal()
    trigger_reject = pyqtSignal()
    def __init__(self,parent=None, LogBrowser=None):
        """
            Setup Default Values.
            Generate Required data's.
        """
        super(MalwareUploadDialog,self).__init__(parent)
        self.setupUi(self)
        self.LogBrowser=LogBrowser
        self.empty_string = ""
        self.NoneString = None
        self.sql_queries = DBManipulator.SqlQueries()
        self.Ratings_Combo.MaxCount = 4
        self.Type_Combo.MaxCount = 2
        self.ratings = ["Great","Good","Low"]
        self.counterRatings = 0
        self.types = ["Local", "Remote"]
        self.counterTypes = 0

        self.UW_buttonBox.accepted.connect(self.successfulValidation)
        self.UW_buttonBox.rejected.connect(self.Reject)
        self.trigger_success.connect(self.Save)
        self.trigger_reject.connect(self.Reject)
        self.trigger_messageBox.connect(self.MessageBox)

        self.Malware_Architectures_Options = {"ALL": self.ALL_Radio,
                                  "x86": self.x86_Radio,
                                  "x64": self.x64_Radio,
                                  "x86_64": self.x86_64_Radio }
        self.platform_name = ""
        self.platform_version = ""
        self.malware_title = ""
        self.malware_description = ""
        self.malware_author = ""
        self.option_name = ""
        self.module_name = ""
        self.arch_name = ""
        self.Malware_JSON_Push.clicked.connect(self.getOption)
        self.Malware_Module_Button.clicked.connect(self.getModule)
        self.filename = None

        for rates in self.ratings: 
            self.counterRatings += 1
            self.Ratings_Combo.addItem(rates)
            self.Ratings_Combo.setCurrentIndex(self.counterRatings)
            self.counterRatings += 1


        for types in self.types:
            self.Type_Combo.addItem(types)
            self.Type_Combo.setCurrentIndex(self.counterTypes)
            self.counterTypes += 1

        for key, value in self.Malware_Architectures_Options.items():
            self.Malware_Architectures_Options[key].pressed.connect(self.setArch)

        self.required_data = {
                            "Name": None,
                            "Version": None, 
                            "Arch": None,
                            "Title": None,
                            "module_name": None,
                            "option_name": None,
                            "Malware_Author": None,
                            "Ratings": None,
                            "Type": None,
                            "Description": None,
                            "UPLOAD_TYPE_ID": "MW_ID"
                            }

    ################################
    #   Input Validation methods
    ################################

    def isPlatformEmpty(self):
        if self.Platform_LineEdit.text() == self.empty_string or self.Platform_LineEdit.text() is self.NoneString:
            self.platform_name = self.Platform_LineEdit.text()
            return True

    def isPlatformVersionEmpty(self):
        if self.Version_Edit.text() == self.empty_string or self.Version_Edit.text() is self.NoneString:
            self.platform_version = self.Version_Edit.text()
            return True
    
    def isTitle_Empty(self):
        """
            @Args: None
            @Default: None
            @Return: True if it is empty.
        """
        if self.Malware_Title_Edit.text() == self.empty_string or self.Malware_Title_Edit.text() is self.NoneString:
            self.malware_title = self.Malware_Title_Edit.text()
            return True

    def isModuleEmpty(self):
        
        if self.module_name == self.empty_string or self.module_name is self.NoneString:
            return True

    def isOptionsEmpty(self):
        if self.option_name == self.empty_string or self.option_name is self.NoneString:
            return True

    def isAuthorEmpty(self):
        if self.Malware_Author_Edit.text() == self.empty_string or self.Malware_Author_Edit.text() is self.NoneString:
            return True

    def isDescriptionEmpty(self):
        """
            @Args: None
            @Default: None
            @Return: True if it is empty.
        """
        if self.Malware_Description_TextEdit.toPlainText() == self.empty_string or self.Malware_Description_TextEdit.toPlainText is self.NoneString:
            self.malware_description = self.Malware_Description_TextEdit.toPlainText()
            return True
            
    def Clean(self):

        self.Platform_LineEdit.setText('')
        self.Version_Edit.setText('')
        self.Malware_Title_Edit.setText('')
        self.Malware_Module_Edit.setText('')
        self.Malware_Author_Edit.setText('')
        self.Malware_Description_TextEdit.setText('')
        self.Malware_Opt_Edit.setText('')
        self.platform_name = ""
        self.platform_version = ""
        self.malware_author = ""
        self.malware_description = ""
        self.malware_title = ""
        self.module_name = ""
        self.option_name = ""

    def Save(self):
        self.Clean()
        CoreLoadConfig.ConfigHandler.AccessWriteLogger("Malware Uploaded %s" % self.module_name,self.LogBrowser)
        self.accept()

    def Reject(self):
        self.Clean()
        CoreLoadConfig.ConfigHandler.ErrorWriteLogger("Malware Upload cancelled!", self.LogBrowser)
        self.reject()

    def setArch(self):
         for key, value in self.Malware_Architectures_Options.items():
            if self.Malware_Architectures_Options[key].isDown():
                self.arch_name = key

    def getSuccessData(self):
        """
            Return the required_data.
        """
        return self.required_data

    def checkFile(self,selected_file):
        """
            Validate the given file. Check for any possible errors.
            @Args: None
            @Default: None
            @return: isFileClean = True
        """
        path_holder = pathlib.Path(selected_file)
        if path_holder.exists():
            if path_holder.is_file():
                if path_holder.stat().st_size == 0 or path_holder.stat().st_size is None:
                    raise CoreException.FileEmptyError("File should not be empty!")
                    return False

                if path_holder.is_symlink():
                    raise CoreException.FileNotSupportedError("Symbolic link not supported")
                    return False
              
               # File Clean if they pass the required identity of file.
                return True

    def MessageBox(self, stringvalue, boxtype="error"):
        """
            Provide Message Box for any types of unknown or invalid inputs.
        """
        messagebox = QMessageBox()
        messagebox.setText(stringvalue)
        messagebox.exec()
        if boxtype == "error":
            CoreLoadConfig.ConfigHandler.ErrorWriteLogger(stringvalue,self.LogBrowser)
        if boxtype == "access":
            CoreLoadConfig.ConfigHandler.AccessWriteLogger(stringvalue,self.LogBrowser)

    ################################
    #   Getter methods
    ################################
    def getAuthor(self):

        if self.isAuthorEmpty():
            return self.empty_string
        self.malware_author = self.Malware_Author_Edit.text()

        return self.malware_author

    def getPlatform(self):
        if self.isPlatformEmpty():
            return self.empty_string
        self.platform_name = self.Platform_LineEdit.text()
    
    def getPlatformVersion(self):
        if self.isPlatformVersionEmpty():
            return self.empty_string
        self.platform_version = self.Version_Edit.text()
        return self.platform_version
        
    def getArch(self):
        """
            @Args: None.)
            @Default: None.
            @Return: Architecture.
        """
        return self.arch_name

    def getTitle(self):
        if self.Malware_Title_Edit.text() == self.empty_string or self.Malware_Title_Edit.text() is self.NoneString:
            return self.empty_string
        self.malware_title = self.Malware_Title_Edit.text()
        return self.malware_title

    def getDescription(self):
        if self.Malware_Description_TextEdit.toPlainText() == self.empty_string or self.Malware_Description_TextEdit.toPlainText() is self.NoneString:
            return self.empty_string

        self.malware_description = self.Malware_Description_TextEdit.toPlainText()
        return self.malware_description

    def getAuthor(self):
        if self.Malware_Author_Edit.text() == self.empty_string or self.Malware_Author_Edit.text() is self.NoneString:
            return self.empty_string

        self.malware_author = self.Malware_Author_Edit.text()
        return self.malware_author
    
    def getRatings(self):
        return self.ratings[self.Ratings_Combo.currentIndex()]

    def getType(self):
        return self.types[self.Type_Combo.currentIndex()]

    def getModule(self):
        self.module_name = QFileDialog.getOpenFileName(self,"Open File","/home")
        if self.module_name != ('',''):
            try:
                self.checkFile(self.module_name[0])
            except CoreException.FileEmptyError as error:
                self.trigger_messageBox.emit("File is empty. Cause: %s" % error)
            except CoreException.FileNotSupportedError as FileNotSupport:
                self.trigger_messageBox.emit("File symbolic link not supported. Cause: %s" % FileNotSupport)
            finally:
                self.Malware_Module_Edit.setText(self.module_name[0])
                self.module_name  = self.module_name[0]
                CoreLoadConfig.ConfigHandler.AccessWriteLogger("Malware module chosen %s" % self.module_name,self.LogBrowser)

    def getModuleName(self):
        return self.module_name

    def getOption(self):
        self.option_name = QFileDialog.getOpenFileName(self,"Open File","/home","*.json")
        if self.option_name != ('',''):
            try:
                self.checkFile(self.option_name[0])
            except CoreException.FileEmptyError as error:
                self.trigger_messageBox.emit("File is empty. Cause: %s" % error)
            except CoreException.FileNotSupportedError as FileNotSupport:
                self.trigger_messageBox.emit("File symbolic link not supported. Cause: %s" % FileNotSupport)
            self.Malware_Opt_Edit.setText(self.option_name[0])
            self.option_name = self.option_name[0]
            CoreLoadConfig.ConfigHandler.AccessWriteLogger("Malware option chosen %s" % self.option_name,self.LogBrowser)

    def getOptionName(self):
        return self.option_name

    def successfulValidation(self):
        if self.getPlatform() == self.empty_string or self.getPlatform is self.NoneString:
            self.trigger_messageBox.emit("Platform should be provided")
            self.reject()
            return False

        self.required_data['Name'] = "\"" + str(self.platform_name) + "\""

        if self.getPlatformVersion() == self.empty_string or self.getPlatformVersion is self.NoneString:
            self.trigger_messageBox.emit("Platform version should be provided")
            self.reject()
            return False

        self.required_data['Version'] = "\"" + str(self.platform_version) + "\""

        if self.getTitle() == self.empty_string or self.getTitle() is None:
            self.trigger_messageBox.emit("Malware Title should be provided")
            self.reject()
            return False

        self.required_data['Title'] = "\"" + str(self.malware_title) + "\""

        if self.getModuleName() == self.empty_string or self.getModuleName() is self.NoneString:
            self.trigger_messageBox.emit("Module should be provided")
            self.reject()
            return False

        if self.getArch() == self.empty_string or self.getArch() is self.NoneString:
            self.trigger_messageBox.emit("Architecture should be chosen")
            self.reject()
            return False

        self.required_data['Arch'] = "\"" + str(self.getArch()) + "\""
        self.required_data['module_name'] = "\"" + str(self.getModuleName()) + "\""

        if self.getOptionName() == self.empty_string or self.getOptionName() is self.NoneString:
            self.trigger_messageBox.emit("Option should be provided")
            self.reject()
            return False

        self.required_data['option_name'] = "\"" +  self.option_name + "\""

        if self.getAuthor() == self.empty_string or self.getAuthor is self.NoneString:
            self.trigger_messageBox.emit("Malware Author should be provided")
            self.reject()
            return False

        self.required_data['Malware_Author'] = "\"" + str(self.malware_author) + "\""

        if self.getDescription() == self.empty_string or self.getDescription() is None:
            self.trigger_messageBox.emit("Description should be provided")
            self.reject()
            return False

        self.required_data['Description'] = "\"" + str(self.malware_description) + "\""
        self.required_data['Ratings'] = "\"" + str(self.getRatings()) + "\""
        self.required_data['Type'] = "\"" + str(self.getType()) + "\""
        self.trigger_success.emit()

    def uploadMalwareClicked(self):
        CoreLoadConfig.ConfigHandler.AccessWriteLogger("Upload Malware Active",self.LogBrowser)
        self.exec()

if __name__ == '__main__':
    app = QApplication(sys.argv)
    upload = ExpUpload()
    upload.UploadExploitClicked()
    app.exec_()
